tests/test_buffered_synced_collection.py

```diff
@@ -0,0 +1,421 @@
+            assert on_disk_dict == raw_dict
+
+        on_disk_dict = self.load(synced_dict)
+        assert 'buffered3' in on_disk_dict
+        assert on_disk_dict == synced_dict
+
+    def test_two_buffered(self, synced_dict, testdata):
+        """Test that a non-buffered copy is not modified."""
+        synced_dict['buffered'] = testdata
+        synced_dict2 = BufferedJSONDict(filename=synced_dict._filename)
+
+        # Check that the non-buffered object is not modified.
+        with synced_dict.buffered():
+            synced_dict['buffered2'] = 1
+            assert 'buffered2' not in synced_dict2
+
+    def test_two_buffered_modify_unbuffered(self, synced_dict, testdata):
+        """Test that in-memory changes raise errors in buffered mode."""
+        synced_dict['buffered'] = testdata
+        synced_dict2 = BufferedJSONDict(filename=synced_dict._filename)
+
+        # Check that the non-buffered object is not modified.
+        with pytest.raises(MetadataError):
+            with synced_dict.buffered():
+                synced_dict['buffered2'] = 1
+                synced_dict2['buffered2'] = 2
+                assert synced_dict['buffered2'] == 1
+                synced_dict['buffered2'] = 3
+                assert synced_dict2['buffered2'] == 2
+                synced_dict2['buffered2'] = 3
+                assert synced_dict['buffered2'] == 3
+
+        self._force_flush_backends()
+
+    def _force_flush_backends(self):
+        # TODO: If client code catches errors raised due to invalid data in the
+        # buffer then attempts to continue, the buffer will be in an invalid
+        # state, and any future attempt to leave buffered mode will trigger
+        # another flush that will error. I'm not sure what the best way to deal
+        # with this problem is. The easiest option is probably to clear the
+        # cache whenever an error occurs so that the buffer is back in a valid
+        # state.
+        from signac.core import buffered_synced_collection
+        for backend in buffered_synced_collection._BUFFERED_BACKENDS:
+            try:
+                backend._cache.clear()
+            except AttributeError:
+                pass
+
+    def test_two_buffered_modify_unbuffered_first(self, synced_dict, testdata):
+        # TODO: What is the expected behavior in this test? Data is only loaded
+        # into the buffer the first time anything happens, so if we enter the
+        # buffered context for one collection then modify an unbuffered
+        # collection within that context _before_ doing anything with the
+        # buffered collection, the buffered version will overwrite it. This
+        # behavior feels slightly unexpected, but I don't know if there's any
+        # way to fix it. For object-local buffering, we could load into the
+        # buffer when entering the context instead of waiting until the first
+        # call to load, but for global buffering there's no equivalent.
```

Comment:
The overwrite is exactly what I expect, based on the implementation. Exiting the context manager should involve writing to disk, and the modification in the middle should be regarded as an invalidation of the assumptions of buffered mode. Allowing `synced_dict` to overwrite `synced_dict2` seems fine, given that we're in the territory of potentially undefined behavior for buffering.

If we wanted to make this "defined and invalid" instead of "undefined," we could store _only_ the file metadata (size and modification time) upon entering the context manager, and wait to load the file contents until accessed. That would let us catch metadata inconsistencies caused by this case of "external modification before buffer load." This would involve splitting the method `_initialize_data_in_cache` into two parts, I believe.

Reply:
@csadorf any thoughts on which approach you prefer? I agree with the statement that this is the expected behavior _based on the implementation_ and that the proposed solution would give something well-defined, so I'm mostly trying to get a sense for what people think a reasonable expectation would be for someone who _doesn't_ know the implementation details.